src/main/scala/exercise2/Stack.scala

```diff
@@ -0,0 +1,29 @@
+package exercise2
+
+import java.util.EmptyStackException
+import scala.annotation.tailrec
+import scala.util.{Failure, Success, Try}
+
+class Stack[T] (elem: Option[T] = None, tail: Option[Stack[T]] = None) extends StackLike[T] {
+
+  override def push(elem: T): StackLike[T] = this.pushOption(Option(elem))
+
+  override def pop(): Try[StackLike[T]] = tail match {
+    case Some(tail) => Success(tail)
+    case None => Failure(new EmptyStackException)
+  }
+
+  override def top(): Option[T] = elem
+
+  override def isEmpty: Boolean = elem.isEmpty && tail.isEmpty
```

Comment:
Why bother checking if `tail` `isEmpty`?

If `tail` is indeed `Some()` but elem is `None`, `isEmpty` would return false yet top would result in a `None`. 

This is a good example of why the `EmptyStack` and `Stack` approach is quite nice, the datastructure itself prevents us from doing that.